File: typescript/token.ts
import { Env } from "./env";

export function preprocessString(env: Env, expr: string): string {
  // works like builtin macro
  // Use a regular expression with global replacement to handle all instances of `(
  expr = expr.replace(/`\(/g, "(quote "); // global search

  // Regular expression to handle different patterns preceded by a backtick
  expr = expr.replace(/`([\w+\-*/!@#$%^&=<>?]+|[^\s\(\)]+)/g, "(quote $1)");

  expr = expr.replace(/\(/g, "( ").replace(/\)/g, " )");

  for (let i = 0; i < env.macros.length; i++) {
    expr = expr.replace(env.macros[i][0], env.macros[i][1]);
  }

  return expr;
}

export function tokenize(env: Env, expr: string): string[] {
  try {
    expr = preprocessString(env, expr);

    const result: string[] = [];

    for (let i = 0; i < expr.length; i++) {
      // skip spaces
      if (expr[i] === " ") {
        continue;
      } else if (expr[i] === "(" || expr[i] === ")") {
        result.push(expr[i]);
      } else if (expr[i] === '"') {
        let token = "";
        result.push('"');

        i++;
        while (expr[i] !== '"') {
          token += expr[i];
          i++;
        }
        result.push(token);
        result.push('"');
      } else if (expr[i] === "{") {
        let token = "";
        result.push("{");

        i++;
        while (expr[i] !== "}") {
          token += expr[i];
          i++;
        }
        result.push(token);
        result.push("}");
      } else if (expr[i] === "'") {
        // '' contains comments
        i++;
        while (1) {
          if (expr[i] === "'") {
            break;
          }
          i++;
        }
      } else {
        let token = "";
        while (i < expr.length && expr[i] !== " " && expr[i] !== ")") {
          token += expr[i];
          i++;
        }
        i--;
        result.push(token);
      }
    }

    return result;
  } catch (error) {
    return [];
  }
}

File: cpp/token.cpp
#include "common.h"
#include <memory>
#include <regex>
#include <string>
#include <vector>

std::string preprocessString(const std::string &expr) {
  std::string result = expr;

  // Use a regular expression with global replacement to handle all instances of
  // `(
  result = std::regex_replace(result, std::regex("`\\("), "(quote ");

  // Regular expression to handle different patterns preceded by a backtick
  result = std::regex_replace(
      result, std::regex("`([\\w+\\-*/!@#$%^&=<>?]+|[^\\s\\(\\)]+)"),
      "(quote $1)");

  return result;
}

std::vector<std::string> tokenize(const std::string &expr) {
  std::string processedLispExpr = preprocessString(expr);
  std::vector<std::string> result;
  size_t length = processedLispExpr.length();

  for (size_t i = 0; i < length; ++i) {
    if (processedLispExpr[i] == ' ') {
      continue;
    } else if (processedLispExpr[i] == '(' || processedLispExpr[i] == ')') {
      result.push_back(std::string(1, processedLispExpr[i]));
    } else if (processedLispExpr[i] == '"') {
      std::string token = "\"";
      ++i;
      while (i < length && processedLispExpr[i] != '"') {
        token += processedLispExpr[i++];
      }
      token += '"';
      result.push_back(token);
    } else if (processedLispExpr[i] == '{') {
      std::string token = "{";
      ++i;
      result.push_back("{");
      while (i < length && processedLispExpr[i] != '}') {
        token += processedLispExpr[i++];
      }
      token += '}';
      result.push_back(token);
      result.push_back("}");
    } else {
      std::string token;
      while (i < length && processedLispExpr[i] != ' ' &&
             processedLispExpr[i] != ')') {
        token += processedLispExpr[i++];
      }
      --i;
      result.push_back(token);
    }
  }

  return result;
}
